File: Problems/python/boj_특정거리의도시찾기.py
#https://www.acmicpc.net/problem/18352
#BFS, Dijkstra

from collections import deque

#BFS 풀이
n, m, k, x = map(int, input().split())
distance = [-1]*(n+1)
distance[x] = 0

graph = [[] for _ in range(n+1)]
for i in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)

q = deque([x])
while q:
    node = q.popleft()
    for i in graph[node]:
        if distance[i] == -1:
            distance[i] = distance[node] + 1
            q.append(i)
            
if k not in distance:
    print(-1)
else:
    for i in range(1, len(distance)):
        if distance[i] == k:
            print(i)
    
#Dijkstra 풀이
import heapq
import sys
input = sys.stdin.readline
INF = int(1e9)

n, e, k, s = map(int, input().split())
graph = [[] for _ in range(n+1)]
dist = [INF]*(n+1)

for i in range(e):
    a, b = map(int, input().split())
    graph[a].append(b)

def dijkstra(start):
    q = []
    dist[start] = 0
    heapq.heappush(q, (0, start))

    while q:
        d, n = heapq.heappop(q)
        if dist[n] < d: continue

        for i in graph[n]:
            cost = d + 1
            if cost < dist[i]:
                dist[i] = cost
                heapq.heappush(q, (cost, i))

dijkstra(s)

result = []
for i in range(1, n+1):
    if dist[i] == k:
        result.append(i)

if result:
    for i in result:
        print(i)
else:
    print(-1)

'''
BFS, Dijkstra로 풀수있는 문제였다.
근데 백준에서 BFS/python3로 채점하면 시간초과 나고 pypy3로 하면 통과가 됐다.
Dijkstra는 그냥 통과가 됐다.
책에서는 모든 간선의 비용이 1이기 때문에 BFS로 풀어도 무방하다고 했다.
'''
    

File: Problems/cpp/boj_특정거리의도시찾기.cpp
//https://www.acmicpc.net/problem/18352
//Dijkstra

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#define INF 1e9

using namespace std;

int n, m, k, s;
vector<vector<int>> graph;
vector<int> dist;

void dijkstra(int start){
    priority_queue<pair<int, int>> pq;
    pq.push(make_pair(0, start));
    dist[start] = 0;

    while (!pq.empty()){
        int d = -pq.top().first;
        int now = pq.top().second;
        pq.pop();

        if (dist[now] < d) continue;

        for (int i=0; i<graph[now].size(); i++){
            int cost = d + 1;
            int next = graph[now][i];
            if (dist[next] > cost){
                dist[next] = cost;
                pq.push(make_pair(-cost, next));
            }
        }
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n >> m >> k >> s;

    graph.resize(n+1);
    dist.resize(n+1, INF);
    
    int a, b;
    for (int i=0; i<m; i++){
        cin >> a >> b;
        graph[a].push_back(b);
    }

    dijkstra(s);

    if (find(dist.begin(), dist.end(), k) != dist.end()){
        for (int i=1; i<n+1; i++){
            if (dist[i]==k)
                cout << i << '\n';
        }
    }
    else
        cout << -1;
}